### PR TITLE
[Improve][Connector-fake] Optimizing Data Generation Strategies #4002

### LINKED ISSUE #4002: [Improve][Connector-fake] Optimizing Data Generation Strategies
### Search before asking

- [X] I had searched in the [feature](https://github.com/apache/incubator-seatunnel/issues?q=is%3Aissue+label%3A%22Feature%22) and found no similar feature requirement.

### Description

At present, the fake first completes the data generation, and then transmits it to the sink, if you test a large amount of data, it will cause memory overflow.  Optimization point：The generated data is directly transmitted to the sink, and it is not necessary to generate all of them.

### Usage Scenario

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by zhilinli123**
## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/incubator-seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by zhilinli123** [comment]
@liugddx Help review

**Comment by zhilinli123** [comment]
issues: https://github.com/apache/incubator-seatunnel/issues/4002

**Comment by TyrantLucifer** [comment]
Thank you for your contribution. Overall looks good to me. Let's waiting CICD

**Comment by TyrantLucifer** [comment]
Unfortunately, your code style has some problems. Give you some tips:
1. Download `checkstyle` plugin in idea and enabled it in the whole project. `checkstyle.xml` you can find it at `tools/checkstyle`
2. Before submit pull request, use the command `mvn checkstyle:check` to check your code style is or not right.

**Comment by zhilinli123** [comment]
I re-debugged it locally
@TyrantLucifer

**Comment by TyrantLucifer** [comment]
> I re-debugged it locally @TyrantLucifer

Waiting CICD.

**Comment by zhilinli123** [comment]
> > I re-debugged it locally @TyrantLucifer
> 
> Waiting CICD.

 my CICD seems to have failed. I don't know what caused it

**Comment by TyrantLucifer** [comment]
> > > I re-debugged it locally @TyrantLucifer
> > 
> > 
> > Waiting CICD.
> 
> my CICD seems to have failed. I don't know what caused it

https://github.com/apache/incubator-seatunnel/actions/runs/4045655710/jobs/6977367847

**Comment by zhilinli123** [comment]
> > > > 我在本地重新调试了@TyrantLucifer
> > > 
> > > 
> > > 等待CICD。
> > 
> > 
> > 我的 CICD 似乎失败了。我不知道是什么原因造成的
> 
> https://github.com/apache/incubator-seatunnel/actions/runs/4045655710/jobs/6977367847

Can you give me some tips to solve the problem? I don't see the underlying problem

**Comment by TyrantLucifer** [comment]
> > > > > 我在本地重新调试了@TyrantLucifer
> > > > 
> > > > 
> > > > 等待CICD。
> > > 
> > > 
> > > 我的 CICD 似乎失败了。我不知道是什么原因造成的
> > 
> > 
> > https://github.com/apache/incubator-seatunnel/actions/runs/4045655710/jobs/6977367847
> 
> Can you give me some tips to solve the problem? I don't see the underlying problem

Case module
image

Case details
image

Case error reason
image

When fake connector defined row, data generation based on this configuration item takes precedence

**Comment by zhilinli123** [comment]
> > > > > > 我在本地重新测试了@TyrantLucifer
> > > > > 
> > > > > 
> > > > > 等候 CICD。
> > > > 
> > > > 
> > > > 我的CICD好像失败了。我不知道是什么原因造成的
> > > 
> > > 
> > > https://github.com/apache/incubator-seatunnel/actions/runs/4045655710/jobs/6977367847
> > 
> > 
> > 你能给我一些解决问题的建议吗？我没有看到潜在的问题
> 
> 案例模块 图片
> 
> 案例详情 图片
> 
> 案例错误原因 图片
> 
> fake connector定义行时，优先根据该配置项生成数据

image
The code in the current project is not clear from where it was loaded

File: seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeDataGenerator.java [hunk 1]
```diff
@@ -92,6 +93,21 @@ public List<SeaTunnelRow> generateFakedRows(int rowNum) {
...
+     **/
+    public void collectFakedRows(int rowNum, Collector<SeaTunnelRow> output) {
+        if (fakeConfig.getFakeRows() != null) {
+            for (FakeConfig.RowData rowData : fakeConfig.getFakeRows()) {
+                output.collect(convertRow(rowData));
+            }
+        }
+        for (int i = 0; i < rowNum; i++) {
+            output.collect(randomRow());
+        }
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeDataGenerator.java]:
```suggestion
        if (fakeConfig.getFakeRows() != null) {
            for (FakeConfig.RowData rowData : fakeConfig.getFakeRows()) {
                output.collect(convertRow(rowData));
            }
        } else {
            for (int i = 0; i < rowNum; i++) {
                output.collect(randomRow());
            }
        }
```

File: seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeSourceReader.java [hunk 2]
```diff
@@ -70,11 +70,8 @@ public void pollNext(Collector<SeaTunnelRow> output) throws InterruptedException
         synchronized (output.getCheckpointLock()) {
             FakeSourceSplit split = splits.poll();
             if (null != split) {
-                // Generate a random number of rows to emit.
-                List<SeaTunnelRow> seaTunnelRows = fakeDataGenerator.generateFakedRows(split.getRowNum());
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeSourceReader.java:74]: remove unused method `generateFakedRows`

Comment (hailin0) [reply seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeSourceReader.java:74]: @zhilinli123 here

Comment (zhilinli123) [reply seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeSourceReader.java:74]: > remove unused method `generateFakedRows`

The generateFakedRows method is used for other test cases

Comment (hailin0) [reply seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeSourceReader.java:74]: @zhilinli123 

update test cases and remove unused method `generateFakedRows`

org.apache.seatunnel.connectors.seatunnel.fake.source.FakeDataGeneratorTest#testRowDataParse
```java
       //List<SeaTunnelRow> seaTunnelRows = fakeDataGenerator.generateFakedRows(fakeConfig.getRowNum());

        List<SeaTunnelRow> seaTunnelRows = new ArrayList<>();
        fakeDataGenerator.collectFakedRows(fakeConfig.getRowNum(), new Collector<SeaTunnelRow>() {
            @Override
            public void collect(SeaTunnelRow record) {
                seaTunnelRows.add(record);
            }

            @Override
            public Object getCheckpointLock() {
                throw new UnsupportedOperationException();
            }
        });
        Assertions.assertIterableEquals(expected, seaTunnelRows);
```

Comment (zhilinli123) [reply seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeSourceReader.java:74]: Thank you for your help. I will submit it again

Comment (TyrantLucifer) [reply seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeSourceReader.java:74]: CICD had been starting. Let's waiting CICD.

Comment (zhilinli123) [reply seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeSourceReader.java:74]: @hailin0 Can you help me locate the reason why my CICD failed

**Comment by liugddx** [comment]
#4061

**Comment by CalvinKirs** [comment]
@liugddx In the case that the contributor is still following up, we'd better help him complete this contribution instead of creating a new PR. This is not a good practice.
CC @Hisoka-X @ashulin 

@zhilinli123  I'm sorry for this. and looking forward to your next contribution.

**Comment by liugddx** [comment]
> @liugddx In the case that the contributor is still following up, we'd better help him complete this contribution instead of creating a new PR. This is not a good practice. CC @Hisoka-X @ashulin
> 
> @zhilinli123 I'm sorry for this. and looking forward to your next contribution.

Excuse me. I originally wanted to add content to this pr, but since I don't have permission, I opened another pr to try.  cc @zhilinli123

**Comment by zhilinli123** [comment]
> @liugddx在贡献者还在跟进的情况下，我们最好帮助他完成这次贡献，而不是创建一个新的 PR。这不是一个好的做法。 CC@Hisoka-X @ashulin
> 
> @zhilinli123 我为此感到抱歉。并期待您的下一次贡献。



> @liugddx在捐献者还在跟进的情况下，我们最好的帮助他完成了这次捐献，而不是创造一个新的公关。这不是一个好的做法。 CC@Hisoka-X @ashulin
> 
> @zhilinli123 我为此感到拥抱。并期待您的下一次奉献。

Thank you for your help and I'd love to contribute

**Pull request closed without merging**